Refactor CMake logic for the CUDA compiler

jngrad opened the pull request:
Description of changes:
- the CUDA compiler guessing mechanism was removed
- the `WITH_CUDA` option is now opt-in
- for non-standard CUDA compilers (Clang, HIP), the user has to provide an extra string option `WITH_CUDA_COMPILER` (any of "nvcc", "clang", "hip")
- if the requested CUDA compiler is not found or is too old, CMake will fail
- the minimal CUDA compiler version numbers are defined in the top-level `CMakeLists.txt`
- the convoluted (and incorrect) CXX optimization flag deduction mechanism in `build_cmake.sh` was replaced with dedicated build types (`Coverage` and `RelWithAssert`)

jngrad committed:
Define C++ version and load CUDA first

Fixes a regression from 1e3fc3b183d: the Boost version depends on
the CUDA version for Intel, therefore CUDA must be loaded first,
but the CUDA version depends on the C++ version, so C++14 must be
defined at the top of the file.

jngrad committed:
Load the CUDA compiler selected by the user

The user can now provide the name of the CUDA compiler as a CMake
flag instead of modifying CMake variables to bypass the guessing
mechanism. CMake fails if the selected compiler is not found. The
FindCUDACompiler helper file was split into separate .cmake files.

jngrad committed:
Check the CUDA version for all compilers

The minimum CUDA version is 9.0 for NVCC and 8.0 for Clang. For the
Clang compiler and the Clang-based HIP compiler, the minimum version
that is tested in CI is 6.0.0.

jngrad committed:
Document CUDA compiler selection mechanism

jngrad committed:
Document how to access CMake variables

jngrad committed:
Remove outdated CMake advice

The ESPResSo CMake project now has proper dependencies for every
target in the source code, documentation and tests. It is no longer
necessary to delete the build directory to start a fresh build.

jngrad committed:
New CUDA compiler selection mechanism in CI

Explicitly toggle GPU code on/off with WITH_CUDA and select the
compiler with WITH_CUDA_COMPILER.

jngrad committed:
Make the default value of WITH_CUDA option OFF

With the CUDA compiler guessing mechanism removed, the user must
explicitly states if CUDA is available and which compiler to use.

jngrad committed:
Patch hipcc environment variables in ROCm 3.X

In ROCm 3.0 and 3.1, environment variables for hipcc and hcc are
overriden by incorrect paths (espressomd/docker#156). This causes
CMake to generate an incorrect linking command for EspressoCore.so:
in `/opt/rocm/bin/hipcc_cmake_linker_helper /opt/rocm -fPIC ...`,
either path `/opt/rocm` is an empty string, or both the linker path
and path `/opt/rocm` are empty strings. Calling `find_package()`
twice with an overriden `HCC_PATH` fixes the linking command.

jngrad committed:
Print CMake version during configuration

Knowing the CMake version is extremely useful when reviewing CMake
logs attached in bug reports.

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -19,7 +19,8 @@
 
 cmake_minimum_required(VERSION 3.10)
-if(${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.13)
+message(STATUS "The CMake identification is ${CMAKE_VERSION}")
```
why not just "CMake version: ${CMAKE_VERSION}"

jngrad replied:
The next line in the CMake log is `-- The C compiler identification is GNU 7.5.0`, but I wouldn't mind using your formulation.

jngrad replied:
formulation was changed

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -286,7 +287,24 @@ set(CMAKE_CXX_EXTENSIONS OFF)
+    list(APPEND CMAKE_MODULE_PATH "${ROCM_HOME}/hip/cmake")
+    find_package(HIP 1.5.18494 MODULE REQUIRED)
+    set(HCC_PATH "${HIP_ROOT_DIR}")
+    find_package(HIP 1.5.18494 MODULE REQUIRED)
```
why do you have to call ```find_package(HIP``` two times?

jngrad replied:
it's explained in 9ac0c09e7

jngrad replied:
I've added an extra CMake comment on top of the git commit explanation.

jngrad commented on cmake/FindCUDACompilerNVCC.cmake:
```diff
@@ -0,0 +1,48 @@
+
+set(CUDA_NVCC_FLAGS_DEBUG "${CUDA_NVCC_FLAGS_DEBUG} -g")
+set(CUDA_NVCC_FLAGS_RELEASE "${CUDA_NVCC_FLAGS_RELEASE} -O3 -DNDEBUG")
+set(CUDA_NVCC_FLAGS_MINSIZEREL "${CUDA_NVCC_FLAGS_MINSIZEREL} -Os -DNDEBUG")
```
several issues here:
* `-Os` is not a valid optimization level: `nvcc fatal   : 's': expected a number`
* there seems to be a difference between `-O2`, `-Xptxas -O2`, `-Xcompiler -O2` ([StackOverflow](https://stackoverflow.com/a/43845984))

jngrad replied:
progress:
* CUDA compiler flags are actually quite complicated. Also, device and host compiler optimization flags don't mix well.
* We don't need to touch the host and device flags in `build_cmake.sh`. The logic is wrong anyway, and it is partially overridden by CMake in Debug builds.

I'm currently removing a lot of useless code in `build_cmake.sh`, but getting the CMake logic right will take me a few more days, I'm afraid. After that, we'll be able to experiment with CMake pseudo-interfaces for GPU targets.

jngrad replied:
I wasted two man hours investigating a Clang compiler error, because when we do
https://github.com/espressomd/espresso/blob/d67ab4aefd263997cbec82015052a19f48b48385/src/core/CMakeLists.txt#L147
we forward CUDA compiler flags to regular .cpp files:
https://github.com/espressomd/espresso/blob/d67ab4aefd263997cbec82015052a19f48b48385/CMakeLists.txt#L165
even though a few lines later there's a mechanism to set those flags to .cu files only:
https://github.com/espressomd/espresso/blob/d67ab4aefd263997cbec82015052a19f48b48385/CMakeLists.txt#L172

jngrad replied:
I wasted one man hour because neither `CMakeLists.txt` nor `build_cmake.sh` throw an error when using the incompatible combination of CMake flags `-DCUDA_NVCC_EXECUTABLE=$(which clang++) -DWITH_COVERAGE=ON`.

jngrad replied:
My last PR introduced a regression: the boost version depends on the CUDA version, so CUDA must be loaded before boost...

jngrad replied:
Why do we override C++ flags only for release builds and not all other types of build???
https://github.com/espressomd/espresso/blob/f291cf448f30b28488830e65556436659870bbf7/src/core/CMakeLists.txt#L78-L79

jngrad replied:
Why `-01` specifically? And why only for release builds?
https://github.com/espressomd/espresso/blob/f291cf448f30b28488830e65556436659870bbf7/CMakeLists.txt#L347-L348

mkuron replied:
FindCUDA didn‘t really have a concept of debug vs. release, so we manually set these flags to match the host compiler and get debuggable or optimized CUDA code as requested.

jngrad replied:
> FindCUDA didn‘t really have a concept of debug vs. release

doesn't it since 3.0? [`CUDA_NVCC_FLAGS_<CONFIG>`](https://cmake.org/cmake/help/v3.0/module/FindCUDA.html?highlight=CUDA_NVCC_FLAGS_%3CCONFIG%3E)

HIP also does: [`HIP_<COMPILER>_FLAGS_<CONFIG>`](https://github.com/ROCm-Developer-Tools/HIP/blob/821c60a3d98f5f4d6cda9440c1452f73edb93b5a/cmake/FindHIP.cmake#L15-L19)

mkuron replied:
I believe that didn't set the flags for both host and device, but that may no longer be true. The `-g -O1` for `WITH_ASAN` is a compromise between good sanitizability and reasonable performance. In Debug mode, you have `-O0 -g` anyway.

jngrad committed:
Modern CMake: naming convention for CUDA variables

The C++ standard used for CUDA code is set in `CMAKE_CUDA_STANDARD`.
Variable `CMAKE_CUDA_VERSION` was renamed to `MINIMAL_CUDA_VERSION`
for clarity.

jngrad committed:
Fix nvcc compiler error

The nvcc `-O<N>` optimization flag can only take a number.

jngrad committed:
Fix clang compiler error

Cannot compile CUDA code with coverage enabled using Clang.

jngrad committed:
Refactor optimization flags, simplify CI script

Rewrite CUDA flags based on the CXX flags:

CMAKE_CXX_FLAGS_DEBUG = -O0 -g
CMAKE_CXX_FLAGS_RELEASE = -O3 -DNDEBUG
CMAKE_CXX_FLAGS_MINSIZEREL = -Os -DNDEBUG
CMAKE_CXX_FLAGS_RELWITHDEBINFO = -O2 -g -DNDEBUG

Add a COVERAGE build type that uses -O0 for host and -O3 for
device. This replaces the logic in the CI script that had to touch
`CMAKE_CXX_FLAGS` and `CUDA_NVCC_FLAGS`. The -O0 optimization flag
for host avoids ending up with source code lines in the gcov output
with neither hit or miss. According to `man gcov`:

> compile your code without optimization if you plan to use gcov
> because the optimization, by combining some lines of code into
> one function, may not give you as much information as you need

jngrad committed:
Remove obsolete CI options

jngrad committed:
Refactor build type and CUDA compiler selection

Generate a warning for incorrect build types and override them with
'Release'. Move the `set(CMAKE_BUILD_TYPE CACHE)` declaration out
of the conditional such that its help message always gets displayed
in cmake/ccmake. List the possible values as properties to allow
cycling in ccmake and cmake-gui (instead of manually typing them).
Same thing for WITH_CUDA_COMPILER. This is achieved by creating a
wrapper around the `option()` function that accepts an enum value
(stored as a string) instead of a boolean value.

jngrad committed:
Update CMake documentation and license headers

jngrad committed:
Add oversubscription guard

The same guard is used in the CMake logic for Python tests.

jngrad committed:
Fix GPU optimization flag for Coverage builds

jngrad commented:
Some jobs are taking longer to compile espresso, yet the test time remains the same. This could be due to caching. One exception is the tutorial job, where the tutorial and sample tests take almost twice longer to run (on the same `coyote11`). On my machine, removing `CUDA_NVCC_FLAGS_DEBUG` didn't change the samples test time, so I'm not sure what the reason might be.

The changes are best visualized by removing the first two commits in the global diff. I'll now work on writing generator expressions in targets instead of writing flags in global variables, and try using target interfaces to reduce code duplication.

jngrad commented:
I can now confirm that the only difference between `python` and 61e02dc8c10 is that `-0g` was added to all C++ files and to the CUDA files processed by Clang. This might explain why compilation takes longer, however the Clang CI job should take less time to run the tests thanks to `-Og`, but now it takes 60% more time to run them. Same thing for tutorials, they should run faster.

There is no difference in line coverage between `-Og` and `-O0` for maxset (unit + integration tests).

KaiSzuttor commented:
any news here?

jngrad commented:
I finally found out the slowdown was caused by an optimization flag `-Og` being injected last in the list of compiler flags, overriding the `-O3` flag (the last optimization flag takes precedence). I now have to figure out which part of the CMake logic is responsible for that.

jngrad commented:
Looks like the C++ flags were always set to `-O3` unless coverage is enabled... The original `build_cmake.sh` is definitely broken.

jngrad commented:
oh come on, `build_cmake.sh` doesn't print the last third of the `$cmake_params` string, I wasted another 2 man hours fixing the wrong compiler flags

jngrad committed:
Test the Release build type in CI

The original `build_cmake.sh` logic was accidentally compiling
ESPResSo with optimization flag -O0 in job "installation", leading
to a significant slowdown in CI.

jngrad committed:
Add a RelWithAssert build type

The "release with assertions" build type restores the original
behavior of the CI pipeline for non-coverage CI jobs: -O3 for
faster tests and -g to get a minimal backtrace in GDB.

jngrad committed:
Print the complete list of CMake options

Move all the if/else logic for CMake options above the function call
that prints shell variables to display the full CMake command in CI.

jngrad committed:
Merge branch 'python' into refactor-cmake-with_gpu

jngrad committed:
Apply clang-format

jngrad committed:
Factor out coverage flags

jngrad commented:
hmm... `set(EspressoCuda_SRC` contains a mix of .cpp and .cu files. We cannot use the solution
```cmake
add_library(gpu_interface INTERFACE)
target_compile_options(gpu_interface INTERFACE "$<$<CONFIG:DEBUG>:-g>")
target_link_libraries(${ARGV0} PUBLIC gpu_interface)
```
because CUDA-specific flags then apply to .cpp files as well, throwing a compiler error. We also can't use `$<$<COMPILE_LANGUAGE:CXX>:-g>` because it's not portable according to the [docs](https://cmake.org/cmake/help/v3.10/manual/cmake-generator-expressions.7.html#manual:cmake-generator-expressions(7)):
> This generator expression has limited use because it is not possible to use it with the Visual Studio generators. Portable buildsystems would not use this expression, and would create separate libraries for each source file language instead:
> ```cmake
> add_library(myapp_c foo.c)
> add_library(myapp_cxx bar.cpp)
> target_compile_options(myapp_cxx PUBLIC -fno-exceptions)
> add_executable(myapp main.cpp)
> target_link_libraries(myapp myapp_c myapp_cxx)
> ```

jngrad commented:
Splitting the `${GPU_TARGET_NAME} = EspressoCore` into two targets with different compiler flags doesn't seem to work even with `PUBLIC`: properties of `EspressoCore` are not inherited by `EspressoCore_cxx` and `EspressoCore_cu`:
```cmake
  add_library(${GPU_TARGET_NAME}_cxx ${${GPU_TARGET_NAME}_sources_cxx})
  add_library(${GPU_TARGET_NAME}_cu  ${${GPU_TARGET_NAME}_sources_cu})
  target_link_libraries(${GPU_TARGET_NAME}_cu PUBLIC gpu_interface)
  add_library(${GPU_TARGET_NAME})
  set_target_properties(${GPU_TARGET_NAME} PROPERTIES LINKER_LANGUAGE "CXX")
  target_link_libraries(${GPU_TARGET_NAME} PUBLIC ${GPU_TARGET_NAME}_cxx ${GPU_TARGET_NAME}_cu)
  target_link_libraries(${GPU_TARGET_NAME} PRIVATE ${CUDA_LIBRARY} ${CUDART_LIBRARY})
```
```
You have called ADD_LIBRARY for library EspressoCore without any source files. This typically indicates a problem with your CMakeLists.txt file
[  1%] Building CXX object src/core/CMakeFiles/EspressoCore_cxx.dir/cells.cpp.o
In file included from /work/jgrad/espresso-fork-PR/src/core/cells.cpp:27:
In file included from /work/jgrad/espresso-fork-PR/src/core/cells.hpp:41:
In file included from /work/jgrad/espresso-fork-PR/src/core/CellStructure.hpp:25:
In file included from /work/jgrad/espresso-fork-PR/src/core/Cell.hpp:22:
/work/jgrad/espresso-fork-PR/src/core/Particle.hpp:22:10: fatal error: 'config.hpp' file not found
#include "config.hpp"
         ^~~~~~~~~~~~
[  3%] Building CXX object src/core/CMakeFiles/EspressoCore_cu.dir/actor/DipolarBarnesHut_cuda.cu.o
/work/jgrad/espresso-fork-PR/src/core/actor/DipolarBarnesHut_cuda.cu:24:10: fatal error: 'cuda_wrapper.hpp' file not found
#include "cuda_wrapper.hpp"
         ^~~~~~~~~~~~~~~~~~
```

jngrad commented:
Oh right, `EspressoConfig` is `PRIVATE`. That explains everything. In fact, we can decouple the CUDA code from the C++ code using `EspressoCore` for C++ and `EspressoCore_cu` for CUDA, and only expose the bare minimum of libraries and header files to `EspressoCore_cu`:
```cmake
target_link_libraries(EspressoCore_cu PRIVATE EspressoConfig shapes)
target_include_directories(EspressoCore_cu PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
target_link_libraries(EspressoCore PRIVATE EspressoCore_cu)
```
This actually compiles on Clang and passes the Python tests.

jngrad commented:
Let me amend the previous statement: we can't separate CUDA from C++ sources files using two targets. If `EspressoCore_cu` is:
- a shared library (58bdbc8d): there's a relink issue: can't install ESPResSo outside the build dir
  ```
  /usr/bin/python3: Relink `/lib/x86_64-linux-gnu/libsystemd.so.0' with `/lib/x86_64-linux-gnu/librt.so.1' for IFUNC symbol `clock_gettime'
  /usr/bin/python3: Relink `/lib/x86_64-linux-gnu/libudev.so.1' with `/lib/x86_64-linux-gnu/librt.so.1' for IFUNC symbol `clock_gettime'
  ```
- a static library (71937be1):
   - Clang fails if `-fPIC` isn't placed judiciously in the list of compiler flags for `EspressoCore`
     ```
     /usr/bin/ld: libEspressoCuda.a(DipolarBarnesHut_cuda.cu.o): relocation R_X86_64_32 against symbol `_Z20initializationKernelv' can not be used when making a shared object; recompile with -fPIC
     ```
   - in addition, there is a linking issue: can't install ESPResSo outside the build dir
     ```
     ImportError: libcudart.so.8.0: cannot open shared object file: No such file or directory
     ```
- an object library (bc89a3e73): can't link `EspressoConfig shapes` nor inherit them: compilation fails

The non-portable `$<COMPILE_LANGUAGE:language>` evaluates to something non-binary: both `$<COMPILE_LANGUAGE:CUDA>` and `$<NOT:$<COMPILE_LANGUAGE:CXX>>` evaluate to false on CUDA files.

I love CMake.

@KaiSzuttor Do you have any suggestion? I can't think of another way to introduce a GPU `INTERFACE` library that would only affect the .cu files and not the .cpp files in target `EspressoCore`.

jngrad commented:
@KaiSzuttor Found the solution: installing thrust 1.9.5 to fix the `va_printf` issue and doing `export LD_LIBRARY_PATH=/work/jgrad/cuda-10.0-thrust195/lib64:$LD_LIBRARY_PATH` to get shared libraries to find the correct CUDA libraries. I'll resume work on this PR.

jngrad committed:
Detect CUDA libraries installed via nvidia-cuda-toolkit

fweik commented on CMakeLists.txt:
```diff
@@ -402,18 +393,21 @@ if(WITH_ASAN AND WITH_MSAN)
 endif()
 
+add_library(cxx_interface INTERFACE)
+target_compile_options(cxx_interface INTERFACE ${cxx_interface_flags})
```
out of currriosity, why do you keep the list in `cxx_interface_flags` instead of directly using `target_compile_options`?

jngrad replied:
We actually need to pass these flags to both the compiler and linker, otherwise we get unresolved symbols (e.g. ASAN in [clang:6.0 logfile](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/223148)).

jngrad replied:
In the line below, I pass the same list to [`target_link_libraries()`](https://cmake.org/cmake/help/v3.12/command/target_link_libraries.html?highlight=target_link#id2), which defines linker flags. CMake 3.13 introduced [`target_link_options()`](https://cmake.org/cmake/help/v3.13/command/target_link_options.html) to pass only linker flags, in an effort to make the intent clearer thanks to the name similarity with `target_compile_options()`.

fweik replied:
Only the `-fsanitize=` no? (c.f.  https://gitlab.icp.uni-stuttgart.de/fweik/mdx/-/blob/master/CMakeLists.txt) Otherwise passing compiler options to the linker doesn't do anything, and is  just confusing. Also you still wouldn't need the list, right?

jngrad replied:
Thanks for the hint! I wasn't sure if `-fsanitize=...` alone was sufficient, but it actually was.

fweik replied:
Yeah, some of the sanitizers have a runtime component, which needs to be linked...

jngrad commented:
The `osx-cuda` CI job is failing because of unresolved symbols in library `EspressoCuda`. These symbols are defined in `EspressoCore`. For example:

```c++
// lb_inertialess_tracers_cuda_interface.hpp
extern IBM_CUDA_ParticleDataInput *IBM_ParticleDataInput_host;
// lb_inertialess_tracers_cuda_interface.cpp
IBM_CUDA_ParticleDataInput *IBM_ParticleDataInput_host = nullptr;
// lb_inertialess_tracers_cuda.cu
    if (IBM_ParticleDataInput_host != NULL) {
```
Did I introduce a design flaw by creating a shared object `EspressoCuda.so` with the cuda code and `EspressoCore.so` for the C++ code? None of the other GPU build seems to mind. If it's not an issue, then we can ignore the CI failure: `osx-cuda` was removed in #3652.

The `intel:19` CI failure is just a regression in the image that wasn't caught in docker CI.

fweik commented:
@jngrad IIRC there are cyclic dependencies between the CUDA and the core code...

jngrad committed:
Modern CMake

Use interface libraries for compiler flags instead of populating
variable with global scope that are injected in the compilation and
linking commands of all libraries (e.g. CMAKE_CXX_FLAGS). Remove
duplicated compile flags. Give meaningful names to variables in
FindCUDA* CMake files. Document CMake policies and add CMP0025 to
distinguish between Clang and AppleClang. Replace simple if/else
blocks by generator expressions.

jngrad commented:
@KaiSzuttor The PR is ready from my side. Let me now if you have additional changes to suggest. Otherwise, I'll rebase commits 58bdbc8dc6854 to 7025e65f to make the git history less chaotic, and resolve the merge conflict. The slowdown issues mentioned 2 weeks ago have been resolved and the `EspressoCuda` and `EspressoCore` are now a single target again to solve the cyclic dependency.

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -19,36 +19,47 @@
 project(ESPResSo)
 include(cmake/FindPythonModule.cmake)
-if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.12")
+include(cmake/option_enum.cmake)
```
if you move line 54 further up, you can omit specifying the whole path. Also you can omit the ```.cmake``` part.

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -130,6 +141,43 @@ endforeach(func_name __PRETTY_FUNCTION__ __FUNCTION__)
+    find_package(CUDA ${MINIMAL_CUDA_VERSION} REQUIRED)
+    find_package(CUDACompilerNVCC ${MINIMAL_CUDA_VERSION} REQUIRED)
+  elseif(WITH_CUDA_COMPILER STREQUAL "hip")
+    set(ROCM_HOME "/opt/rocm" CACHE FILEPATH "Path to AMD ROCm")
+    list(APPEND CMAKE_MODULE_PATH "${ROCM_HOME}/hip/cmake")
+    find_package(HIP 1.5.18494 MODULE REQUIRED)
+    # patch HCC_PATH environment variable and reload HIP
+    if(HIP_VERSION VERSION_LESS "3.1")
+      set(HCC_PATH "${HIP_ROOT_DIR}")
+    else()
+      set(HCC_PATH "${ROCM_HOME}/hcc")
+    endif()
+    find_package(HIP 1.5.18494 MODULE REQUIRED)
+    find_package(CUDACompilerHIP 6.0 REQUIRED)
```
can't we hide this in a module?

jngrad replied:
yes, I'll move it to `cmake/FindCUDACompilerHIP.cmake`

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -130,6 +141,43 @@ endforeach(func_name __PRETTY_FUNCTION__ __FUNCTION__)
+    if(WITH_COVERAGE)
+      message(
+        FATAL_ERROR
+          "The combination of options '-DWITH_CUDA_COMPILER=${WITH_CUDA_COMPILER} -DWITH_COVERAGE=${WITH_COVERAGE}' is not supported."
```
I think having the name for the option hardcoded in the error string is not maintainable

jngrad replied:
fixed

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -130,6 +141,43 @@ endforeach(func_name __PRETTY_FUNCTION__ __FUNCTION__)
+  else()
+    message(
+      FATAL_ERROR
+        "Cannot process option -DWITH_CUDA_COMPILER='${WITH_CUDA_COMPILER}'")
```
see above

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -283,70 +331,56 @@ set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/${PYTHON_INSTDIR}/espressomd")
+    -Wno-unused-parameter
+    -Wno-missing-braces
+    $<$<AND:$<CXX_COMPILER_ID:GNU>,$<STREQUAL:${CMAKE_SYSTEM_PROCESSOR},arm>>:-Wno-psabi>
+    $<$<AND:$<NOT:$<CXX_COMPILER_ID:Intel>>,$<NOT:$<AND:$<CXX_COMPILER_ID:GNU>,$<VERSION_LESS:$<CXX_COMPILER_VERSION>,7.0.0>>>>:-Wno-implicit-fallthrough>
```
hm... this is really hard to read.

jngrad replied:
just these two, or the following ones too?

jngrad replied:
I'm reverting the ones with more than 1 AND/OR, here and in other CMake files.

KaiSzuttor reviewed (dismissed):
Looks good to me besides the comments. Thanks @jngrad that looks much cleaner now.

mkuron commented on CMakeLists.txt:
```diff
@@ -130,6 +141,43 @@ endforeach(func_name __PRETTY_FUNCTION__ __FUNCTION__)
+      set(HCC_PATH "${ROCM_HOME}/hcc")
+    endif()
+    find_package(HIP 1.5.18494 MODULE REQUIRED)
+    find_package(CUDACompilerHIP 6.0 REQUIRED)
```
Where does that `6.0` version number come from? Is that the Clang version of hipcc? I don't think it makes sense to require a minimum version there because the HIP version already implies a certain minimum Clang version.

jngrad replied:
It's the Clang version. I put 6.0 here since this is the oldest Clang version we're testing. Do you know where I could find a table showing the HIP <=> Clang version mapping?

jngrad replied:
I've removed that Clang check in be61c3a1e.

jngrad committed:
Simplify HIP detection

Move ROCm path patching logic in FindCUDACompilerHIP.cmake and
check only the HIP version.

jngrad committed:
Fix Travis shell error

The double quotes were not removed by the shell interpreter.

jngrad committed:
Merge branch 'python' into refactor-cmake-with_gpu

jngrad committed:
Correctly detect CUDA versions with Clang

Clang returns "version unknown" for unsupported CUDA libraries,
or doesn't return a version string (depending on CMAKE_CXX_FLAGS),
causing the CMake regex to store the complete Clang stdout in the
CUDA_VERSION variable instead of a valid version number. This is
now fixed, and the CUDA version is now shown as <major>.<minor>.
